### PR TITLE
Kubernetes: initial CRD framework

### PR DESCRIPTION
**Pull request by mclarke47**
## Hey, I just made a Pull Request!

* Start framework for CRD support
* Add support for Argo Rollouts - [Rollout canary spec][1]
* Refactor how Kubernetes Objects data is accessed by components

Screenshot 2021-03-09 at 14 10 42

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

[1]: https://argoproj.github.io/argo-rollouts/features/canary/

### CODE REVIEW

File: plugins/kubernetes/src/components/CustomResources/ArgoRollouts/Rollout.tsx
Comment (adamdmharvey) [inline plugins/kubernetes/src/components/CustomResources/ArgoRollouts/Rollout.tsx:34]: Not sure but I think this may apply here?

https://backstage.io/docs/architecture-decisions/adrs-adr010

Comment (mclarke47) [reply plugins/kubernetes/src/components/CustomResources/ArgoRollouts/Rollout.tsx:34]: Thanks, I'll take a look!

Comment (mclarke47) [reply plugins/kubernetes/src/components/CustomResources/ArgoRollouts/Rollout.tsx]: Sound be fixed

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4892-d7gz2tex9qy8sbftyyyvqy2tyyogxw9f.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/604788dbf285af0497a46d86

File: plugins/kubernetes-backend/schema.d.ts
Comment (freben) [inline plugins/kubernetes-backend/schema.d.ts:25]: Should this one be optional perhaps?

Comment (mclarke47) [reply plugins/kubernetes-backend/schema.d.ts]: done!

Comment (mclarke47) [reply plugins/kubernetes-backend/schema.d.ts]: done!

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4892-d7gz2tex9qy8sbftyyyvqy2tyyogxw9f.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/604788dbf285af0497a46d86

**Comment by mclarke47** [comment]
This should be ready now

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4892-d7gz2tex9qy8sbftyyyvqy2tyyogxw9f.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/604788dbf285af0497a46d86

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4892-d7gz2tex9qy8sbftyyyvqy2tyyogxw9f.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/604788dbf285af0497a46d86

File: docs/features/kubernetes/configuration.md
Comment (benjdlambert) [inline docs/features/kubernetes/configuration.md:134]: Just a quick thought for this, can we not get the plural form by describing the `CRD` with the API?

File: plugins/kubernetes/package.json
Comment (benjdlambert) [inline plugins/kubernetes/package.json:37]: Don't think you'll need to depend on this right? You should be able to grab it from `core` itself. I don't think anything should depend on `core-api` :)

File: docs/features/kubernetes/configuration.md
Comment (mclarke47) [reply docs/features/kubernetes/configuration.md:134]: I think this is a great idea and  was actually recommended by @nora99 also. I think we can add it later as a feature.

File: plugins/kubernetes/package.json
Comment (mclarke47) [reply plugins/kubernetes/package.json:37]: Will replace

Comment (mclarke47) [reply plugins/kubernetes/package.json]: done

File: docs/features/kubernetes/configuration.md
Comment (benjdlambert) [reply docs/features/kubernetes/configuration.md:134]: Cool - let's iterate!

**Review by benjdlambert — APPROVED** [review]

**Comment by benjdlambert** [comment]
Looks good - without digging into the code too much :)

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4892-d7gz2tex9qy8sbftyyyvqy2tyyogxw9f.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/604788dbf285af0497a46d86

**Comment by benjdlambert** [comment]
Apologies @mclarke47 this is my fault for not saying that I also wanted @freben and @Rugvip to take a look over this too to make sure that there was nothing else that standing out before merging, gonna hold off on the release until they've had a flick through too. Sorry!

**Comment by tszngai** [comment]
Hi all I was look through the documents and couldn't find anything related to this update. I'm wondering if there's any configurations I need to do to surface Rollout objects in the dashboard? I have added the proper labels, made sure service account has read access to rollouts and the latest version of plugin is used. But still rollout is not showing up. Thanks!

**Comment by tszngai** [comment]
> Hi all I was look through the documents and couldn't find anything related to this update. I'm wondering if there's any configurations I need to do to surface Rollout objects in the dashboard? I have added the proper labels, made sure service account has read access to rollouts and the latest version of plugin is used. But still rollout is not showing up. Thanks!

Ah nvm I found it, need to update `kubernetes.customResources` in app-config.yaml

**Pull request merged**
